_posts/2018-10-08-self-arrow.md

```diff
@@ -0,0 +1,108 @@
+---
+layout: post
+title: "Desambiguating 'this' in Scala, or what the hell means `self =>`?"
+author: ragb
+---
+
+This is a quick Scala tip on a somehow unknown feature of the Scala language, what I call *self aliasing* or *unconstrained self type annotations*. One sees that in some libraries or even in the Scala standard library itself.
+It goes something like this:
+
+```scala
+trait MyTrait { self =>
+  //...
+}
+```
+
+What? You probably have seen a similar self type annotation within the context of the now infamous cake pattern:
+
+```scala
+trait A {
+  def doAStuff = ???
+}
+
+trait B {
+  this: A =>
+  def doBStuff = doAStuff
+}
+```
+
+This means that the `B` trait can only be mixed in something that mixes `A`, which makes `A` members available to `B`.
+Someone just thought this pattern was awesome for dependency injection and everyone started using that, but it doesn't scale at all... And then lots of pain and legacy to maintain.
+
+Anyway, getting back to our first example, why that `self =>`
+ thing? It's not constraining anything whatsoever... It is giving another name to `this`:
+ 
+```scala
+trait MyTrait {self => }
+// is equivalent to
+trait MyTrait2 {private val self = this}
+```
+
+Without any context it can be difficult to grasp why one needs to give a different name to `this`, so let us show a small example.
+
+We will show the definition for a function (`Function1`) from some type `A` to `Option[B]` (for some type `B`). Moreover we will implement the `
+andThen` method to compose two such functions, if the first function returns some value `B`
+(there is a powerful abstraction for this concept called a [kleisli arrow][kleisli], but it's out of the scope of this article).
+
+Let's start from the basic trait's definition:
+
+```scala
+trait Function1Option[-A, +B] {
+  def apply(a: A): Option[B]
+}
+```
+
+So we define an abstract `apply` method (as function-like things do in scala) which concrete instances must implement. Still, without knowing about the concrete `apply`
+one can easily define the `andThen` implementation, can't we?
```

Comment:
> one can easily define the `andThen` implementation, can't we?

I searched how tag questions are used when the subject is "one", but I couldn't find an answer that would seem trustworthy. I'm inclined to say the second clause would be ", can't it?" but it still sounds strange. So in my opinion the second clause should be replaced with ", right?" or ", correct?".